File: src/main/scala/org/nygenome/als/graphdb/value/GeneOntology.scala
package org.nygenome.als.graphdb.value

import org.nygenome.als.graphdb.util.StringUtils

case class GeneOntology (
                          goTermAccession:String,
                          goDomain:String,
                          goName:String,
                          goDefinition:String
                        ){

}
object GeneOntology  extends ValueTrait {
  // method to parse a gene ontology entry from the uniprot file
  def parseGeneOntologyEntry(aspect: String, entry:String):GeneOntology ={
    val tuple2:Tuple2[String,String]  = StringUtils.parseGeneOntologyEntry(entry)
    GeneOntology(tuple2._2, aspect, tuple2._1," ")
  }
}

File: src/main/kotlin/org/biodatagraphdb/alsdb/model/GeneOntology.kt
package org.biodatagraphdb.alsdb.model

data class GeneOntology(
        val goTermAccession: String,
        val goDomain: String,
        val goName: String,
        val goDefinition: String = " "
) {

    companion object : AlsdbModel {
        private fun resolveGeneOntologyTuple(goEntry: String): Pair<String, String> {
            val index = goEntry.indexOf('[') + 1
            val index2 = 0.coerceAtLeast(index - 1)
            return Pair(goEntry.substring(0, index2).trim(),
                    goEntry.slice(index..index + 9))
        }

        fun parseGeneOntologyEntry(aspect: String, entry: String): GeneOntology {
            val tuple2: Pair<String, String> = resolveGeneOntologyTuple(entry)
            return GeneOntology(goTermAccession = tuple2.second, goDomain = aspect,
                    goName = tuple2.first)

        }
    }

}